feat: Show wallet balance

usame-algan opened the pull request:
## What it solves

Displays the connected wallet balance in the header and in the tx-flow

[Figma Designs](https://www.figma.com/file/ptTs6lDBeUuLNySroJ5PiF/Web-Master-File?type=design&node-id=8647-11732&mode=design&t=25lotpzWMdjN7X2q-0) kudos to @TanyaEfremova !

## How this PR fixes it

- extracts the `useWallet` hook into a global context

## How to test it

1. Observe the wallet balance showing in the header for any connected wallet
2. Observe that it always reflects the connected wallet network not necessarily the network the safe is on
3. Create a transaction and observe seeing the wallet balance in the Execute block
4. Observe the new design according to the linked Figma

## Screenshots
<img width="602" alt="Screenshot 2023-10-23 at 09 20 39" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/12a4e16e-532d-4ed2-9740-97c05afe33e9">
<img width="697" alt="Screenshot 2023-10-23 at 09 20 58" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/f67499e7-4ecd-43da-ae61-8d5f82a0c547">
<img width="338" alt="Screenshot 2023-12-04 at 09 48 42" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/4e78435a-94db-4028-9137-d21fb0bf4054">
<img width="335" alt="Screenshot 2023-12-04 at 09 49 25" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/ffda8b44-d1e2-4c7d-a4de-e9824d5edb2b">
<img width="700" alt="Screenshot 2023-12-07 at 14 09 40" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/320b4c17-9951-4457-a177-1318cd436537">

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

katspaugh commented on src/components/common/WalletBalance/index.tsx:
```diff
@@ -0,0 +1,21 @@
+
+  return (
+    <>
+      {formatVisualAmount(balance, currentChain?.nativeCurrency.decimals ?? 18)}{' '}
+      {currentChain?.nativeCurrency.symbol ?? 'ETH'}
```
Could you extrace 18 and ETH into a `const NATIVE_TOKEN_PLACEHLDER`?

katspaugh commented on src/hooks/wallets/useWalletBalance.ts:
```diff
@@ -0,0 +1,19 @@
+import { useWeb3ReadOnly } from '@/hooks/wallets/web3'
+import { type BigNumber } from 'ethers'
+
+const useWalletBalance = () => {
```
Could you add a return type here?

katspaugh commented on src/components/common/WalletInfo/styles.module.css:
```diff
@@ -8,6 +8,37 @@
+.walletIcon {
+  position: absolute;
+  display: flex;
+  -webkit-box-pack: center;
+  justify-content: center;
+  -webkit-box-align: center;
```
What are these -webpack rules? Possible to stick to standard CSS?

usame-algan replied:
Oh looks like a copy-paste error. Will remove them

katspaugh reviewed (commented):
Nice little feature!

usame-algan commented on src/hooks/wallets/useWalletBalance.ts:
```diff
@@ -0,0 +1,19 @@
+      return undefined
+    }
+
+    return web3ReadOnly.getBalance(wallet.address, 'latest')
```
Unfortunately, the balance that onboard returns is already formatted as a string. We could use this in the header even though it might have a different format than our number formatting rules but @schmanu made a valid point that if we want to warn the user that their balance is not enough to execute a transaction the amount in WEI or similar raw value would be better.

usame-algan commented:
Note: I suggest we implement the network selector logos in a separate PR since this PR is already quite big

schmanu reviewed (approved):
LGTM!

francovenica commented:
So this balance showing is fine, but if the network doesn't have relayer it will never be displayed.

Is expected to show this only for networks with relayer?
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/ee492a25-c634-4119-89e3-77720006c844)

francovenica commented:
The rest is fine.

I tried with MM chrome extension, MM phone app, Trezor, Ledger, Safe-as-owner and in all of them it looks good
The balance updates everytime you open the display in the top right or refresh the form in the tx

I wasn't able to test in social login since it's blocked in this PR

katspaugh commented:
> So this balance showing is fine, but if the network doesn't have relayer it will never be displayed.

IMO, we should always show this block. Obviously, without the radio buttons if no relayer.

katspaugh commented on src/components/tx/BalanceInfo/index.tsx:
```diff
@@ -0,0 +1,28 @@
+      <div>
+        <Stack direction="row" spacing={0.5} alignItems="center">
+          <Typography variant="body2" fontWeight={700} letterSpacing="0.1px">
+            Wallet balance
```
`Connected wallet balance` perhaps? To not confuse it with the Safe's own ETH balance.

usame-algan commented:
I changed it to always show the `Connected wallet balance` block now either with or without relayer. Added a screenshot in the PR description. @francovenica can you test this new case? Thanks!

usame-algan commented:
> > On the top right corner, is confusing that the balance shown is from the network the safe is on, but the "Network" label shows the network the wallet is currently connected to.
> 
> @francovenica The reasoning for this is that we only switch the network to the Safe's network when the user submits / executes a transaction. If we would show the balance in the wallet's network it could seem like you have enough balance to execute the transaction although it's just a balance on an unrelated network. I agree that it is a bit confusing though. Probably we could add a warning Icon or something.

I remember we talked about prompting the user to switch network when they connect their wallet instead of delaying it to when they try to sign/execute a transaction. Thats how other dapps do it and even go as far as forcing the user to switch before enabling the UI (e.g. CowSwap).

francovenica commented:
Ok, thanks for the response.

I get now that is more important to show the user if he has funds to do tx in the current network than to show how much he has in whatever network he's connected to. Still a bit confusing but it makes sense.

The fix to show the balance even without the relayer looks good
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/d5e30202-48bc-456e-8082-7f21329a5e92)

LGTM
